nemo/collections/asr/models/wav2vec/wav2vec_model.py

```diff
@@ -0,0 +1,541 @@
+        self.project_q = None
+        self.target_glu = None
+        self.final_proj = None
+
+
+class ConvFeatureEncoder(nn.Module):
+    """
+        Converts input raw audio into features for downstream transformer model.
+        Uses 1D convolutional blocks with GeLU activation.
+    """
+
+    def __init__(
+        self,
+        conv_layers: List[Tuple[int, int, int]],
+        mode: Wav2VecConvExtractorMode = Wav2VecConvExtractorMode.default,
+        conv_bias: bool = False,
+    ):
+        super().__init__()
+
+        def block(
+            n_in, n_out, k, stride, is_layer_norm=False, is_group_norm=False, conv_bias=False,
+        ):
+            def make_conv():
+                conv = nn.Conv1d(n_in, n_out, k, stride=stride, bias=conv_bias)
+                nn.init.kaiming_normal_(conv.weight)
+                return conv
+
+            assert (is_layer_norm and is_group_norm) is False, "layer norm and group norm are exclusive"
+
+            if is_layer_norm:
+                return nn.Sequential(
+                    make_conv(),
+                    nn.Sequential(TransposeLast(), nn.LayerNorm(dim, elementwise_affine=True), TransposeLast()),
+                    nn.GELU(),
+                )
+            elif is_group_norm:
+                return nn.Sequential(make_conv(), nn.GroupNorm(dim, dim, affine=True), nn.GELU(),)
+            else:
+                return nn.Sequential(make_conv(), nn.GELU())
+
+        in_d = 1
+        self.conv_layers = nn.ModuleList()
+        for i, cl in enumerate(conv_layers):
+            assert len(cl) == 3, "invalid conv definition: " + str(cl)
+            (dim, k, stride) = cl
+
+            self.conv_layers.append(
+                block(
+                    in_d,
+                    dim,
+                    k,
+                    stride,
+                    is_layer_norm=mode is Wav2VecConvExtractorMode.layer_norm,
+                    is_group_norm=mode is Wav2VecConvExtractorMode.default and i == 0,
+                    conv_bias=conv_bias,
+                )
+            )
+            in_d = dim
+
+    def forward(self, x):
+        # BxT -> BxCxT
+        x = x.unsqueeze(1)
+        for conv in self.conv_layers:
+            x = conv(x)
+        return x
+
+
+class Wav2VecTransformerEncoder(nn.Module):
+    def __init__(self, cfg: Wav2VecTransformerConfig):
+        super().__init__()
+
+        conv_cfg = cfg.conv
+
+        self.dropout = cfg.dropout
+        self.embedding_dim = cfg.encoder.embedding_dim
+        self.layer_norm_first = cfg.encoder.layer_norm_first
+
+        # positional convolutional embeddings
+        self.pos_conv = nn.Conv1d(
+            self.embedding_dim,
+            self.embedding_dim,
+            kernel_size=conv_cfg.conv_pos,
+            padding=conv_cfg.conv_pos // 2,
+            groups=conv_cfg.conv_pos_groups,
+        )
+        dropout = 0
+        std = math.sqrt((4 * (1.0 - dropout)) / (conv_cfg.conv_pos * self.embedding_dim))
```

Comment:
Do these initializations come from the fairseq codebase?

Reply:
Yes they do!